ledger/src/blockstore_db.rs

```diff
@@ -1014,7 +1131,53 @@ impl<'a> WriteBatch<'a> {
     }
 }
 
-fn get_cf_options(access_type: &AccessType) -> Options {
+struct PurgedSlotFilter<C: Column + ColumnName>(Slot, CString, PhantomData<C>);
+
+impl<C: Column + ColumnName> CompactionFilter for PurgedSlotFilter<C> {
+    fn filter(&mut self, _level: u32, key: &[u8], _value: &[u8]) -> CompactionDecision {
+        use rocksdb::CompactionDecision::*;
+
+        let slot_in_key = C::slot(C::index(key));
+        if slot_in_key >= self.0 {
+            Keep
+        } else {
+            Remove
+        }
+    }
```

Comment:
Hmm, how does this play out for the TransactionStatus and AddressSignature columns? Those do not get pruned on the same cadence as the other columns; they retain data until an entire range is below the lowest_cleanup_slot, and remove all at once. What are the perf impacts of running compaction over data that has not been purged?

Reply:
@CriesofCarrots oh, thanks for chiming in. Yeah, I need to read and think about these columns to decide what to do. I'm wondering if possible we can simplify the speical puring logic at `LedgerCleanupService` if we don't need strict purge timing consistency? Alternatively, new mechanism can ignore those special column families.

> What are the perf impacts of running compaction over data that has not been purged?

I think it will be some wasted rocksdb's low-priority background thread cpu time and io time; not ideal but better than current full-range compaction happening now unintendedly. I hope this my comments help you to understand the current situation and my solution to it.
https://github.com/solana-labs/solana/pull/16697#issuecomment-825662256
https://github.com/solana-labs/solana/pull/16697/commits/1fe713e9a053876b9ed15760299738aaf0168a2d

Reply:
Do we still need to call `DeleteRange` given this compaction filter? From the docs https://github.com/facebook/rocksdb/wiki/Compaction-Filter:  

>RocksDB provides a way to delete or modify key/value pairs based on custom logic in background. It is handy for implementing custom garbage collection, *like removing expired keys based on TTL*, or dropping a range of keys in the background."

Also this makes it sound like the compaction filter leaves behind a tombstone, but might not reclaim the memory?

>If it request to filter the key, the value is replaced by a deletion marker. Note that if the output level of the compaction is the bottom level, no deletion marker will need to be output."  

From these *oudated* docs (it says outdated and replaced with `DeleteRange`), it seems like if we use this method, we don't need the newer recommended `DeleteRange`? https://github.com/facebook/rocksdb/wiki/Delete-A-Range-Of-Keys:

 >apply compaction filter together with CompactRange(). You can write a compaction filter that can filter out keys from deleted range. When you want to delete keys from a range, call CompactRange() for the range to delete. While the compaction finishes, the keys will be dropped. "

Reply:
For the TransactionStatus and AddressSignature we only had a custom deletion logic because the keys were not prefixed by `slot` (they needed to be searchable by signature) so  it didn't fit the `DeleteRange` mold. If we switch to this new compaction filter, where we can explicitly grab the key based on different column families like `let slot_in_key = C::slot(C::index(key));` above, then I think we can fold that logic into this pattern as well!

Reply:
pinging @ryoqun for thoughts on removing `DeleteRange` in case you missed this 😃

Reply:
pinging @ryoqun for thoughts on removing `DeleteRange` in case you missed this 😃

Reply:
@carllin Yeah, I'll take a look for handling of these special columns today, which has been mysterious to me up to now... ;) not mentioned on this thread, but I think we also should be careful for the handling of bigtable fallback mechanism like so.

Reply:
@ryoqun cool! hmmm maybe the special handling for those columns can be avoided if the `DeleteRange` logic is removed. I'm not entirely sure how bigtable interacts with the above proposal, @CriesofCarrots may know better.

Reply:
@CriesofCarrots @carllin After some thoughts, I think we can stop `delete_range()`-ing for `TransactionStatus` and `AddressSignature`. Could you check my following thoughts and new commits?



> Also quick note, this "deletion marker" makes it sound like the compaction filter leaves behind a tombstone, but might not reclaim the memory (might be good to double check), but maybe the compaction takes care of this.

the deletion markers from compaction filters should be handled very similarly to the normal tombstones by `delete()`. So, I think the memory is reclaimed correctly. (I'm seeing no memory leak from test nodes).



> From these _oudated_ docs (it says outdated and replaced with `DeleteRange`), it seems like if we use this method, we don't need the newer recommended `DeleteRange`? https://github.com/facebook/rocksdb/wiki/Delete-A-Range-Of-Keys:

As said above, I don't think we need to call `DeleteRange` by principle, assuming we use compaction_filter. `DeleteRange` sounds promising but it has following concerns for our usecase (so I want to avoid using it):

1.  `DeleteRange` can only used to purge based on `primary_index`, not `slot`.
1. Then, it means we'll need the primary index switching logic, meaning back to square one
1. `DeleteRange` would affect very huge number of keys (100M or more) with primary_index. So, this could cause spikes and some bad performance implications? (I'm testing this with nodes now, informally)



> for the `TransactionStatus` and `AddressSignature` columns we only had a custom deletion logic because the keys were not prefixed by `slot` (they needed to be searchable by signature) so it didn't fit the `DeleteRange` mold. If we switch to this new compaction filter, where we can explicitly grab the key based on different column families like `let slot_in_key = C::slot(C::index(key));` above, then I think we can fold that logic into this pattern as well!

Yeah, I think we can realize like this: 20c25a2faf53107bdd03d8acc087369a1c41bdd0 b826ec24c816cd9b9dfe208036c1bda82734bf8f

20c25a2faf53107bdd03d8acc087369a1c41bdd0: I tried to minimize the code change. So I introduced noop new enum variant
b826ec24c816cd9b9dfe208036c1bda82734bf8f: for consistent view, I added these changes. (hope source comments explain things well). This is a bit like a dirty hack. But overall this is good call over the alternative or status quo. The downside of this solution is that code touching those special column families must be aware of the async filtering compaction. But I think this is acceptable because the code surface is very limited. The alternative solution would start increment primary_index (2, 3, 4, etc) then compaction_filter expires keys based on primary_index. But, the alternative incurs more implementation change and makes ledger binary incompatible also this doesn't amortize the storage cleaning IOs well.

FYI, I ditched my half-done attempt to introduce some type safety for this read consistency by introducing `AliveSlot` because it was too much tedious with no super good merits. Also, I avoided custom iterator which is aware of `lowest_cleanup_slot`. These solutions seem overkill and uncalled abstraction... xD

Also, I noticed we have a race condition for some rpc reads regardless of my pr, because we don't lockout LedgerCleanupService while the rpc is reading some history of account over multiple BlockStore calls. But I must confess I never bothered to fix it properly, assuming this is very very minor issue. (Ideally, we must hold a read lock for  `blockstore.lowest_cleanup_slot` whole time for processing a single rpc reqeust, but we release it in the middle. Also, simply holding it like that would easily starve the `LederCleanupService` for busy rpc nodes).

In the distant future, I think we can remove the implementation code altogether and even save 8 bytes from every keys in those CFs and reduce the lookup cost by half without prefixing primary_index....

> I think we also should be careful for the handling of bigtable fallback mechanism like so.

After some proper code read, I concluded these my code changes shouldn't confuse the fallback mechanism code because rpc's `get_signature_status` just always fallback to `bigtable_ledger_storage` if given requested signature is missing in the rocksdb.

Reply:
Hey @ryoqun! Catching up with your work here. I'll have a better response on the actual code after I look at it 😆 . However, I wanted to confirm your Bigtable conclusions...
The RPC service queries Bigtable if the requested data cannot be found in Blockstore, and (in the case of blocks) it cannot be shown to have been skipped by the cluster. As long as `TransactionStatus` and `AddressSignature` data is kept and consistent through at least the minimum slot in the other columns, your changes should be noop as far as Bigtable is concerned.

Reply:
Okay, I think I'm caught up now. Just to clarify: the potential inconsistency in the special columns with this method is that they may retain data for slots that have been deleted from other columns, correct? If so, this isn't actually much different that current behavior, except the lingering data will probably be a lot less.
Seems like a great change.

> In the distant future, I think we can remove the implementation code altogether and even save 8 bytes from every keys in those CFs and reduce the lookup cost by half without prefixing primary_index....

That will be great! I can take on that clean up. Depending on the backward compatibility of those changes, I would also like to change the key for `AddressSignatures` at that time: from `(u64, Pubkey, Slot, Signature)` to `(Pubkey, Slot, u64, Signature)`, where the new `u64` is the transaction's index in the block, enabling `getSignaturesForAddress` to return all results in chronological order. (Bigtable does this now, but Blockstore can't.)

Reply:
> As long as TransactionStatus and AddressSignature data is kept and consistent through at least the minimum slot in the other columns, your changes should be noop as far as Bigtable is concerned.

Cool. thanks for confirming.

> Just to clarify: the potential inconsistency in the special columns with this method is that they may retain data for slots that have been deleted from other columns, correct? ... 

yeah, this again is correct.

Also, the following new mode of consistency problem would show up from _raw_ accesses because of no delete_range for these special CFs and the very nature of async compaction with filters:

Specifically, keys older than `lowest_cleanup_slot` can disappear at any moment and at any order for each signtures and addresses and combination of both. This consistency problem can be easily solved by lower bound like this: 
b826ec24c816cd9b9dfe208036c1bda82734bf8f And I think this guard is sufficient.

> That will be great! I can take on that clean up.

I'm very happy to pass the future work to you. :)

Reply:
> I think this guard is sufficient.

Yep, I agree. Plus it's something we probably should have been doing already